File: IOSApp/IOSApp/Presentation/ViewModels/RecetaViewModel.swift
//
//  RecetaViewModel.swift
//  IOSApp
//
//  Created by Irene Otero Díaz on 6/7/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import sharedApp

class RecetaViewModel: ObservableObject {
    
    private let caseUses: UseCases
    private let receta: Receta
    
    @Published var state: RecetaState = RecetaState()
    
    init(
        caseUses: UseCases,
        receta: Receta
    ){
        self.caseUses = caseUses
        self.receta = receta
        
        self.reprintReceta()
    }
    
    func onTriggerEvent(event: RecetaEventos) -> Void {
        switch event{
            case is RecetaEventos.onaAddIngrediente:
                print("Intentando añadir ingrediente " +
                      (event as! RecetaEventos.onaAddIngrediente).nombre +
                      "de la receta " + String(Int(self.state.cestaCompra_id)) +
                      "de la lista de recetas: " + String(Int(self.state.receta_id))
                )
                addIngrediente(
                    ingrediente: Alimento(
                        id_cestaCompra: KotlinInt(integerLiteral: Int(self.state.cestaCompra_id)),
                        id_receta: KotlinInt(integerLiteral: Int(self.state.receta_id)),
                        id_alimento: nil,
                        nombre: (event as! RecetaEventos.onaAddIngrediente).nombre,
                        cantidad: (event as! RecetaEventos.onaAddIngrediente).cantidad,
                        tipoUnidad: (event as! RecetaEventos.onaAddIngrediente).tipoUnidad,
                        active: true
                    )
                )
            case is RecetaEventos.onDeleteIngrediente:
                deleteIngrediente(id_ingrediente: (event as! RecetaEventos.onDeleteIngrediente).id_ingrediente)
                
            default:
                print("Evento de receta no conocido")
        }
    }
    
    private func addIngrediente(ingrediente: Alimento) -> Void {
        caseUses.addIngredienteReceta.addIngredienteReceta(
            id_cestaCompra: Int32(truncating: ingrediente.id_cestaCompra!),
            id_receta: Int32(truncating: ingrediente.id_receta!),
            nombre: ingrediente.nombre,
            cantidad: ingrediente.cantidad,
            tipoUnidad: ingrediente.tipoUnidad
        ).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                if dataState?.data != nil {
                    self.reprintReceta()
                }
            }
        )
    }
    
    private func deleteIngrediente(id_ingrediente: Int32) -> Void {
        caseUses.deleteIngredienteReceta.deleteIngredientereceta(id_ingrediente: id_ingrediente).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                if dataState?.data != nil {
                    self.reprintReceta()
                }
            }
        )
    }
    
    private func reprintReceta() -> Void {
        let currentState:RecetaState = self.state.copy() as! RecetaState
        if(receta.id_cestaCompra != -1 || receta.id_Receta != -1){
            print("El id de la receta es: " + String(Int(truncating: self.receta.id_Receta!)))
            caseUses.getDatosReceta.getDatosReceta(id_receta: self.receta.id_Receta as! Int32).collectFlow(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState?.data != nil {
                        self.state = self.state.doCopy(
                            cestaCompra_id: self.receta.id_cestaCompra,
                            receta_id: dataState?.data?.id_Receta as! Int32,
                            nombre: self.receta.nombre,
                            cantidad: String(Int(self.receta.cantidad)),
                            imagen: dataState?.data?.imagenSource ?? "",
                            ingredientes: dataState?.data?.ingredientes ?? [],
                            queueError: currentState.queueError,
                            resultadoAutoComplete: currentState.resultadoAutoComplete
                        )
                    }
                }
            )
        }
    }
}

File: AndroidApp/src/main/java/com/otero/recipetoshop/android/presentation/controllers/cestacompra/recetas/RecetaViewModel.kt
package com.otero.recipetoshop.android.presentation.controllers.cestacompra.recetas

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.otero.recipetoshop.Interactors.Common.ActualizarAutoComplete
import com.otero.recipetoshop.Interactors.cestascompra.recetas.AddIngredienteReceta
import com.otero.recipetoshop.Interactors.cestascompra.recetas.DeleteIngredienteReceta
import com.otero.recipetoshop.Interactors.cestascompra.recetas.EditIngrediente
import com.otero.recipetoshop.Interactors.cestascompra.recetas.GetDatosReceta
import com.otero.recipetoshop.domain.util.TipoUnidad
import com.otero.recipetoshop.events.cestacompra.receta.RecetaEventos
import com.otero.recipetoshop.presentationlogic.states.recetas.RecetaState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class RecetaViewModel
@Inject
constructor(
    savedStateHandle: SavedStateHandle,
    private val addIngredienteReceta: AddIngredienteReceta,
    private val getDatosReceta: GetDatosReceta,
    private val deleteIngredienteReceta: DeleteIngredienteReceta,
    private val actualizarAutoComplete: ActualizarAutoComplete,
    private val editIngrediente: EditIngrediente
): ViewModel() {
    val recetaState = mutableStateOf(RecetaState())

    init {
        try {
            savedStateHandle.get<Int>("recetaid")?.let{ recetaid ->
                viewModelScope.launch {
                    recetaState.value = recetaState.value.copy(receta_id = recetaid)
                    if (recetaState.value.receta_id == -1){
                        throw Exception("ID receta == -1")
                    }
                }
            }
            savedStateHandle.get<Int>("cestacompraid")?.let{ cestacompraid ->
                viewModelScope.launch {
                    recetaState.value = recetaState.value.copy(cestaCompra_id = cestacompraid)
                    if (recetaState.value.receta_id == -1){
                        throw Exception("ID cestacompraid == -1")
                    }
                }
            }
            //Pinto los ingredientes cogidos.
            reprintContenidoReceta(id_receta = recetaState.value.receta_id)
        } catch (e: Exception){
            println("No se ha podido obtener el identificador de la lista de recetas: " + e.message)
        }
    }

    fun onTriggerEvent(event: RecetaEventos){
        when(event){
            is RecetaEventos.onAutocompleteRecetaChange -> {
                val response = actualizarAutoComplete.actualizarAutoComplete(query = event.query)
                recetaState.value = recetaState.value.copy(resultadoAutoComplete = response)
            }
            is RecetaEventos.onClickAutocompleteReceta -> {
                recetaState.value = recetaState.value.copy(resultadoAutoComplete = emptyList())
            }
            is RecetaEventos.onaAddIngrediente -> {
                addAlimentoReceta(nombre = event.nombre, cantidad = event.cantidad, tipoUnidad = event.tipoUnidad)
            }
            is RecetaEventos.onDeleteIngrediente -> {
                deleteIngredienteReceta(id_ingrediente = event.id_ingrediente)
            }
            is RecetaEventos.onEditIngrediente -> {
                editarIngrediente(
                    id_alimento = event.id_ingrediente,
                    nombre = event.nombre,
                    cantidad = event.cantidad,
                    tipo = event.tipoUnidad
                )
            }
            else -> {
                throw Exception("ERROR")
            }
        }
    }

    private fun editarIngrediente(id_alimento: Int, nombre: String, cantidad: Int, tipo: String) {
        editIngrediente.editIngrediente(
            id_receta = recetaState.value.receta_id,
            id_cestaCompra = recetaState.value.cestaCompra_id,
            id_alimento = id_alimento,
            nombre = nombre,
            cantidad = cantidad,
            tipoUnidad = TipoUnidad.parseTipoUnidad(tipo)
        ).onEach { dataState ->
            //Actualizo la receta.
            reprintContenidoReceta(recetaState.value.receta_id)
        }.launchIn(viewModelScope)
    }

    private fun deleteIngredienteReceta(id_ingrediente: Int) {
        deleteIngredienteReceta.deleteIngredientereceta(
            id_ingrediente = id_ingrediente
        ).onEach { dataState ->
            dataState.data?.let {
                //Reseteo la lista de ingredientes con el nuevo ingrediente añadido.
                reprintContenidoReceta(id_receta = recetaState.value.receta_id)
            }
        }.launchIn(viewModelScope)
    }

    private fun addAlimentoReceta(nombre: String, cantidad: Int, tipoUnidad: TipoUnidad) {
        addIngredienteReceta.addIngredienteReceta(
            id_cestaCompra = recetaState.value.cestaCompra_id,
            id_receta = recetaState.value.receta_id,
            nombre = nombre,
            cantidad = cantidad,
            tipoUnidad = tipoUnidad
        ).onEach { dataState ->
            dataState.data?.let {
                //Reseteo la lista de ingredientes con el nuevo ingrediente añadido.
                reprintContenidoReceta(id_receta = recetaState.value.receta_id)
            }
        }.launchIn(viewModelScope)
    }

    private fun reprintContenidoReceta(id_receta: Int) {
        //Reseteo la lista de ingredientes.
        recetaState.value = recetaState.value.copy(ingredientes = listOf())

        //Obtengo los ingredientes de la Base de datos renovados.
        getDatosReceta.getDatosReceta(id_receta = id_receta).onEach { dataState ->
            dataState.data?.let { receta ->
                //Los refresco de nuevo.
                recetaState.value = recetaState.value.copy(
                    cestaCompra_id = receta.id_cestaCompra,
                    receta_id = receta.id_Receta!!,
                    nombre = receta.nombre,
                    ingredientes = receta.ingredientes,
                    cantidad = receta.cantidad.toString(),
                    imagen = if(receta.imagenSource == null) "" else receta.imagenSource!!,
                )
            }
        }.launchIn(viewModelScope)
    }

}